packages/compass-components/src/hooks/use-stacked-component.tsx

```diff
@@ -0,0 +1,62 @@
+import type { ComponentType, ForwardedRef } from 'react';
+import React, { createContext, useContext, useMemo } from 'react';
+import { cx, css } from '@leafygreen-ui/emotion';
+
+type StackedComponentProviderProps = {
+  zIndex?: number;
+};
+
+const StackedComponentContext =
+  createContext<StackedComponentProviderProps | null>(null);
+
+export const StackedComponentProvider = ({
+  zIndex,
+  children,
+}: StackedComponentProviderProps & { children: React.ReactNode }) => {
+  return (
+    <StackedComponentContext.Provider value={{ zIndex }}>
+      {children}
+    </StackedComponentContext.Provider>
+  );
+};
+
+export const useStackedComponent = () => {
+  return useContext(StackedComponentContext);
+};
+
+// TODO: LG-4109. This should be eventually supported by the LG design system
+export const withStackedComponentStyles = function <
+  ComponentProps extends { className?: string }
+>(
+  WrappedComponent: ComponentType<ComponentProps>
+): ComponentType<ComponentProps> {
+  const ComponentWithStackedStyles = (
+    props: ComponentProps,
+    ref: ForwardedRef<ComponentType<ComponentProps>>
+  ) => {
+    const context = useStackedComponent();
+    const stackedElementStyles = useMemo(() => {
+      if (context?.zIndex) {
+        return css({
+          zIndex: context.zIndex,
+        });
+      }
+      return '';
+    }, [context]);
+    return (
+      <WrappedComponent
+        ref={ref}
+        {...props}
+        className={cx(props.className, stackedElementStyles)}
```

Comment:
Just wondering, most (if not all) of the components that we need to add this for support popoverZIndex prop. Why are we not overriding this and rely on className merging? What if I actually pass popover zIndex to the component, what will take precedence?

Reply:
I flipped the two and applied className takes precedence now. 

What do you mean by "Just wondering, most (if not all) of the components that we need to add this for support popoverZIndex prop."? I think i am missing something here.

Reply:
So in leafygreen every component that renders in a portal like this (see docs for [Menu](https://www.mongodb.design/component/menu/documentation/), [Popover](https://www.mongodb.design/component/popover/documentation/), [Combobox](https://www.mongodb.design/component/combobox/documentation/), [Tooltip](https://www.mongodb.design/component/tooltip/documentation/), [Inline Definition](https://www.mongodb.design/component/inline-definition/documentation/) etc.) all have a `popoverZIndex` property that is used to override the z-index of a container. In some of those components that's more or less the only way to actually be able to pass a z-index value to the container that renders it, like in the inline definition case. I was thinking that this would probably be the way to pass the z-index value from context to the components, or at least it needs to be accounted for when rendering so that your value here doesn't override it (right now we have a few components that use it and pass a pretty high z-index value)

Reply:
I can see `popoverZIndex` documented for [Popover](https://www.mongodb.design/component/popover/documentation/) but not for [Modal](https://www.mongodb.design/component/modal/documentation/). Are there other places we plan on using this?

Reply:
I just added that in [f48926c](https://github.com/mongodb-js/compass/pull/5588/commits/f48926cc12f02704edff6a848a46ee211209d39e)